flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/FlinkSplitPlanner.java

```diff
@@ -137,6 +121,60 @@ static CloseableIterable<CombinedScanTask> planTasks(
     }
   }
 
+  private static Long resolveToSnapshotId(Table table, ScanContext context) {
+    if (context.endTag() != null) {
+      Preconditions.checkArgument(
+          table.snapshot(context.endTag()) != null,
+          "Cannot find snapshot with tag %s",
+          context.endTag());
+      return table.snapshot(context.endTag()).snapshotId();
+    }
+
+    if (context.endSnapshotId() != null) {
+      return context.endSnapshotId();
+    }
+
+    if (context.endSnapshotTimestamp() != null) {
+      return SnapshotUtil.snapshotIdAsOfTime(table, context.endSnapshotTimestamp());
+    }
+    return null;
+  }
+
+  private static Long resolveStartSnapshotId(Table table, ScanContext context) {
+    if (context.startTag() != null) {
+      Preconditions.checkArgument(
+          table.snapshot(context.startTag()) != null,
+          "Cannot find snapshot with tag %s",
+          context.startTag());
+      return table.snapshot(context.startTag()).snapshotId();
+    }
+
+    if (context.startSnapshotId() != null) {
+      return context.startSnapshotId();
+    }
+
+    if (context.startSnapshotTimestamp() != null) {
+      return getStartSnapshotId(table, context.startSnapshotTimestamp());
+    }
+
+    return null;
+  }
+
+  private static Long getStartSnapshotId(Table table, Long startTimestamp) {
+    Snapshot oldestSnapshotAfter = SnapshotUtil.oldestAncestorAfter(table, startTimestamp);
+    Preconditions.checkArgument(
+        oldestSnapshotAfter != null,
+        "Cannot find a snapshot older than %s for table %s",
+        startTimestamp,
+        table.name());
+
+    if (oldestSnapshotAfter.timestampMillis() == startTimestamp) {
+      return oldestSnapshotAfter.snapshotId();
+    } else {
+      return oldestSnapshotAfter.parentId();
```

Comment:
Could this be `null`?

Reply:
`oldestSnapshotAfter.parentId()` may be `null`. For `null`, we will not use it, see line 94:
```
      Long startSnapshotId = resolveStartSnapshotId(table, context);
      if (startSnapshotId != null) {
        scan = scan.fromSnapshotExclusive(startSnapshotId);
      }
```